src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs

```diff
@@ -38,87 +39,93 @@ public interface IConcurrencyManager
       /// <param name="numberOfCoresToUse">Number of cores to use. Use 0 or negative to take all cores</param>
-      /// <param name="cancellationToken">Cancellation token to cancel the threads</param>
       /// <param name="data">List of data to consume by the workers</param>
-      /// <param name="action">A function to run on each worker on each piece of data</param>
+      /// <param name="idFunc"></param>
+      /// <param name="action">
+      ///    A function running on each worker on each piece of data. It is assume that the action runs on its
+      ///    own thread
+      /// </param>
+      /// <param name="cancellationToken">Cancellation token to cancel the threads</param>
       /// <returns>Dictionary binding a result for each input data after running the action on it</returns>
-      Task<IReadOnlyDictionary<TData, ConcurrencyManagerResult<TResult>>> RunAsync<TData, TResult>
-      (
-         int numberOfCoresToUse, 
-         CancellationToken cancellationToken, 
-         IReadOnlyList<TData> data, 
-         Func<TData, string> id,
-         Func<int, CancellationToken, TData, Task<TResult>> action
-      ) where TResult : class;
+      Task<IReadOnlyDictionary<TData, ConcurrencyManagerResult<TResult>>> RunAsync<TData, TResult>(
+         int numberOfCoresToUse,
+         IReadOnlyList<TData> data,
+         Func<TData, string> idFunc,
+         Func<int, TData, CancellationToken, Task<TResult>> action,
+         CancellationToken cancellationToken
+      );
    }
 
    public class ConcurrencyManager : IConcurrencyManager
    {
       private readonly int _maximumNumberOfCoresToUse = Math.Max(1, Environment.ProcessorCount - 1);
+
       public async Task<IReadOnlyDictionary<TData, ConcurrencyManagerResult<TResult>>> RunAsync<TData, TResult>
-      (
-         int numberOfCoresToUse, 
-         CancellationToken cancellationToken, 
-         IReadOnlyList<TData> data, 
-         Func<TData, string> id,
-         Func<int, CancellationToken, TData, Task<TResult>> action
-      ) where TResult : class
+      (int numberOfCoresToUse,
+         IReadOnlyList<TData> data,
+         Func<TData, string> idFunc,
+         Func<int, TData, CancellationToken, Task<TResult>> action,
+         CancellationToken cancellationToken)
       {
          if (numberOfCoresToUse <= 0)
             numberOfCoresToUse = _maximumNumberOfCoresToUse;
+
          var concurrentData = new ConcurrentQueue<TData>(data);
          numberOfCoresToUse = Math.Min(numberOfCoresToUse, concurrentData.Count);
 
          var results = new ConcurrentDictionary<TData, ConcurrencyManagerResult<TResult>>();
-         //Starts one task per core
-         var tasks = Enumerable.Range(0, numberOfCoresToUse).Select(async coreIndex =>
+
+         //Splits the action based in the number of cores available. 
+         //No thread will be created here. If the actions are all running on the same thread, the effect of the concurrency execution will be inexistent
```

Comment:
added some comments to explain that we do not run stuff concurrently here in fact... NAme of the class is a bit misleading